From 96b2383fe431cb697f4f9aaf1314982ae26bcbe7 Mon Sep 17 00:00:00 2001
From: Natanael Copa <ncopa@alpinelinux.org>
Date: Thu, 4 May 2023 13:12:35 +0200
Subject: [PATCH 1001/1006] alpine: add meta-data network-interfaces and
 resolv_conf

Allow user to set a network-interfaces file in meta-data and set up dns
servers via resolv_conf.

This is for compatibility with cloud-init, and is a simple way to let
user configure static ip or advanced network

ref: https://cloudinit.readthedocs.io/en/latest/reference/datasources/nocloud.html#example-meta-data
ref: https://cloudinit.readthedocs.io/en/latest/reference/modules.html#resolv-conf
---
 lib/tiny-cloud/cloud/alpine/imds |  1 +
 lib/tiny-cloud/cloud/alpine/init | 50 ++++++++++++++++++++++++++++++++
 tests/tiny-cloud-alpine.test     | 41 ++++++++++++++++++++++++++
 3 files changed, 92 insertions(+)
 create mode 120000 lib/tiny-cloud/cloud/alpine/imds
 create mode 100644 lib/tiny-cloud/cloud/alpine/init
 create mode 100755 tests/tiny-cloud-alpine.test

diff --git a/lib/tiny-cloud/cloud/alpine/imds b/lib/tiny-cloud/cloud/alpine/imds
new file mode 120000
index 0000000..104b5ca
--- /dev/null
+++ b/lib/tiny-cloud/cloud/alpine/imds
@@ -0,0 +1 @@
+../nocloud/imds
\ No newline at end of file
diff --git a/lib/tiny-cloud/cloud/alpine/init b/lib/tiny-cloud/cloud/alpine/init
new file mode 100644
index 0000000..2435c75
--- /dev/null
+++ b/lib/tiny-cloud/cloud/alpine/init
@@ -0,0 +1,50 @@
+# Tiny Cloud - Init Functions
+# vim:set ts=4 et ft=sh:
+
+# usage: replace_word <search> <replace> <list>...
+replace_word() {
+    local search="$1" replace="$2"
+    shift 2
+    for word in "$@"; do
+        if [ "$word" = "$search" ]; then
+            echo "$replace"
+        else
+            echo "$word"
+        fi
+    done
+}
+
+INIT_ACTIONS_EARLY="$(replace_word set_default_interfaces set_network_interfaces $INIT_ACTIONS_EARLY)"
+
+add_once() {
+    local file="$1"
+    shift
+    for line; do
+        if ! grep -x -F "$line" "$file" 2>/dev/null; then
+            mkdir -p "${file%/*}"
+            printf "%s\n" "$line" >> "$file"
+        fi
+    done
+}
+
+set_resolv_conf() {
+    # resolv.conf
+    local nameservers="$(imds meta-data/resolv_conf/nameservers)"
+    for i in $nameservers; do
+        local server="$(imds meta-data/resolv_conf/nameservers/$i)"
+        add_once "$ROOT"/etc/resolv.conf "nameserver $server"
+    done
+}
+
+init__set_network_interfaces() {
+    local interfaces="$(imds meta-data/network-interfaces)"
+    mkdir -p "$ROOT"/etc/network
+    if [ -n "$interfaces" ]; then
+        printf "%s\n" "$interfaces" > "$ROOT"/etc/network/interfaces
+    elif ! [ -f "$ROOT"/etc/network/interfaces ]; then
+        init__set_default_interfaces
+    fi
+    if ! grep -q dhcp "$ROOT"/etc/network/interfaces; then
+        set_resolv_conf
+    fi
+}
diff --git a/tests/tiny-cloud-alpine.test b/tests/tiny-cloud-alpine.test
new file mode 100755
index 0000000..ba98448
--- /dev/null
+++ b/tests/tiny-cloud-alpine.test
@@ -0,0 +1,41 @@
+#!/usr/bin/env atf-sh
+
+. $(atf_get_srcdir)/test_env.sh
+
+export PREFIX="$srcdir"
+export MOCK=echo
+export CLOUD=alpine
+lib="$srcdir"/lib/tiny-cloud/cloud/alpine/init
+
+init_tests \
+	set_network_config_network_interfaces
+
+set_network_config_network_interfaces_body() {
+	fake_metadata_nocloud <<-EOF
+		network-interfaces: |
+		  auto eth1
+		  iface eth1
+		    address 192.168.100.1
+		    netmask 255.255.255.0
+
+		resolv_conf:
+		  nameservers:
+		    - 8.8.8.8
+		    - 8.8.4.4
+	EOF
+
+	atf_check \
+		-o match:"rc-update" \
+		-e match:"set_network_interfaces .*DONE" \
+		tiny-cloud early
+	atf_check \
+		-o match:"auto eth1" \
+		-o match:"iface eth1" \
+		-o match:"address 192.168.100.1" \
+		cat etc/network/interfaces
+
+	atf_check \
+		-o match:"^nameserver 8.8.8.8$" \
+		-o match:"^nameserver 8.8.4.4$" \
+		cat etc/resolv.conf
+}
-- 
2.40.1

